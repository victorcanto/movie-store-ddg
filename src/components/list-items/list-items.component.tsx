import React, { useEffect, useState } from 'react';
import {
	Box,
	List,
	Button,
	Grid,
	Typography,
	IconButton,
	Collapse,
} from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import { TransitionGroup } from 'react-transition-group';
import {
	CartItem,
	FavoriteItem,
	removeFavoriteItem,
	removeCartItem,
	addCartItem,
	Movie,
} from '../../shared';

import {
	selectCartItems,
	selectFavoriteItems,
	selectIsCartOpen,
	selectIsFavoritesOpen,
	setCartItems,
	setFavoriteItems,
	useAppDispatch,
	useAppSelector,
} from '../../store';
import { ListItem } from '../list-item/list-item.component';

export interface ListItemsProps {
	title: string;
	emptyMessage: string;
	onCloseMenuHandler: () => void;
}

export function ListItems({
	title,
	emptyMessage,
	onCloseMenuHandler,
}: ListItemsProps) {
	const [listItems, setListItems] = useState<CartItem[] | FavoriteItem[]>([]);
	const isCartOpen = useAppSelector(selectIsCartOpen);
	const isFavoritesOpen = useAppSelector(selectIsFavoritesOpen);
	const cartItems = useAppSelector(selectCartItems);
	const favoriteItems = useAppSelector(selectFavoriteItems);
	const dispatch = useAppDispatch();

	const removeItemHander = (item: CartItem | FavoriteItem) => () => {
		if (isCartOpen) {
			const newCartItems = removeCartItem(cartItems, item);
			if (newCartItems) {
				dispatch(setCartItems(newCartItems));
				return;
			}
		}

		if (isFavoritesOpen) {
			const newFavoriteItems = removeFavoriteItem(favoriteItems, item);
			if (newFavoriteItems) {
				dispatch(setFavoriteItems(newFavoriteItems));
			}
		}
	};

	const addItemToCartHandler = (item: CartItem) => () => {
		const newCartItems = addCartItem(cartItems, item as Movie);

		if (newCartItems) {
			dispatch(setCartItems(newCartItems));
		}
	};

	const clearItemHandler = () => {
		if (isCartOpen) {
			dispatch(setCartItems([]));
			return;
		}

		if (isFavoritesOpen) {
			dispatch(setFavoriteItems([]));
		}
	};

	useEffect(() => {
		if (isCartOpen) {
			setListItems(cartItems);
			return;
		}
		if (isFavoritesOpen) {
			setListItems(favoriteItems);
		}
	}, [isCartOpen, isFavoritesOpen, cartItems, favoriteItems]);

	return (
		<Box sx={{ flexGrow: 1, maxWidth: 752 }}>
			<Grid container spacing={2}>
				<Grid item xs={12} md={12}>
					<Box>
						<Box
							sx={{
								display: 'flex',
								justifyContent: 'space-between',
								alignItems: 'center',
							}}
						>
							<Typography
								sx={{ mt: 4, mb: 2 }}
								variant='h4'
								component='div'
							>
								{title}
							</Typography>

							<IconButton
								size='large'
								aria-label='close'
								color='inherit'
								onClick={onCloseMenuHandler}
							>
								<CloseIcon />
							</IconButton>
						</Box>

						{listItems.length ? (
							<Button onClick={clearItemHandler}>Esvaziar</Button>
						) : null}
					</Box>

					<List data-testid='list-items'>
						<TransitionGroup>
							{listItems.length ? (
								listItems.map((item) => {
									return (
										<Collapse key={item.id}>
											<ListItem
												item={item}
												removeItem={removeItemHander(
													item,
												)}
												addItemToCart={addItemToCartHandler(
													item,
												)}
											/>
										</Collapse>
									);
								})
							) : (
								<Collapse>
									<Typography component='p'>
										{emptyMessage}
									</Typography>
								</Collapse>
							)}
						</TransitionGroup>
					</List>
				</Grid>
			</Grid>
		</Box>
	);
}
