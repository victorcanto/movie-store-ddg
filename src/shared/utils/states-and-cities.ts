import statesAndCities from '../data/estados-cidades.json';

export const getBrazilianStates = () =>
	statesAndCities.estados.map(({ nome }) => nome);

export const getBrazilianCitiesByState = (stateName: string) =>
	statesAndCities.estados.find(({ nome }) => nome === stateName)?.cidades;
