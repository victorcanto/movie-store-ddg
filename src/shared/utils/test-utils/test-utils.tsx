import React, { PropsWithChildren } from 'react';
import { render as rtlRender } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { setupStore, AppStore, RootState } from '../../../store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}

export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {},
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return (
			<Provider store={store}>
				<BrowserRouter>{children}</BrowserRouter>
			</Provider>
		);
	}
	return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export { renderWithProviders as render };
