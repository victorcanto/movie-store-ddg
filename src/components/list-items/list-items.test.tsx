import React from 'react';
import { render } from '../../shared/utils/test-utils';
import { ListItems, ListItemsProps } from './list-items.component';

const onCloseMenuHandlerMock = jest.fn();

const propsMock: ListItemsProps = {
	emptyMessage: 'empty test',
	title: 'test',
	onCloseMenuHandler: onCloseMenuHandlerMock,
};

test('snapshot of the base state', () => {
	const { asFragment } = render(<ListItems {...propsMock} />);
	expect(asFragment()).toMatchSnapshot();
});

test('should have in the document', () => {
	const { getByTestId } = render(<ListItems {...propsMock} />);
	expect(getByTestId('list-items')).toBeInTheDocument();
});
