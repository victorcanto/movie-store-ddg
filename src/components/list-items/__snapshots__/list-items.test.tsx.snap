// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-399ykn"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-69i1ev"
          >
            <div
              class="MuiTypography-root MuiTypography-h4 css-1mtqh3w-MuiTypography-root"
            >
              test
            </div>
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
          data-testid="list-items"
        >
          <div>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                  >
                    empty test
                  </p>
                </div>
              </div>
            </div>
          </div>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
